ansible_base/resource_registry/registry.py

```diff
@@ -0,0 +1,131 @@
+                            self.model_map[label] = {}
+
+                        if hasattr(view, "action"):
+                            action = view.action
+                            if action == "partial_update":
+                                action = "update"
+                            if action not in self.model_map[label]:
+                                self.model_map[label][action] = []
+                            self.model_map[label][action].append((method, path))
+                except:  # noqa E722
+                    pass
+
+
+class ResourceConfig:
+    model_label = None
+    model = None
+    externally_managed = None
+    managed_serializer = None
+    parent_resources = None
+    actions = None
+    name_field = None
+
+    def __init__(self, model, shared_resource: SharedResource = None, parent_resources: List[ParentResource] = None, name_field: str = None):
+        model = get_concrete_model(model)
+        self.model_label = model._meta.label
+
+        managed_serializer = None
+        externally_managed = False
+        if name_field is None:
+            name_field = "name"
+
+        if shared_resource:
+            managed_serializer = shared_resource.serializer
+            externally_managed = not shared_resource.is_provider
+
+        parent_map = {}
+        if parent_resources:
+            for parent in parent_resources:
+                parent_map[parent.model._meta.label] = parent
+
+        self.model = model
+        self.externally_managed = externally_managed
+        self.managed_serializer = managed_serializer
+        self.parent_resources = parent_map
+        self.actions = ResourceInspector().model_map.get(self.model_label, {})
+        self.name_field = name_field
+
+
+class ResourceRegistry:
+    registry = {}
+
+    def __init__(self, resource_list: List[ResourceConfig], service_api_config: ServiceAPIConfig = None):
+        self._validate_api_config(service_api_config)
+        self.api_config = service_api_config
+        for r in resource_list:
+            self.registry[r.model_label] = r
+
+    def _validate_api_config(self, config):
+        """
+        Needs to validate that:
+            - Viewsets have the correct serializer, pagination and filter classes
+            - Service type is set to one of awx, galaxy, eda or gateway
+        """
+        assert config.service_type in ["aap", "awx", "galaxy", "eda"]
+
+    def get_resources(self):
+        return self.registry
+
+    def get_config_for_model(self, model=None, model_label=None) -> ResourceConfig:
+        if model:
+            return self.registry[model._meta.label]
+        if model_label:
+            return self.registry[model_label]
+
+        raise AttributeError("Must include either model or model_label arg.")
+
+
+def get_registry() -> ResourceRegistry:
+    from django.conf import settings
+
+    if hasattr(settings, "ANSIBLE_BASE_RESOURCE_CONFIG_MODULE"):
```

Comment:
I've seen a lot of Django apps implement a registry of some sort, but I've never seen this. I don't quite get it, you're using the apps `.ready` method as a hook to do finalization/setup of the registry. Why does this need the project using it to give an import location with this setting? Instead of having an expected value in that module like `RESOURCE_LIST`, can't you just pass whatever value that is to your registry?

Reply:
So looking back over the code, one of the things that made me really nervous about having a gobal `ResourceRegistry` object that the app imports (similar to how you have `PermissionRegistry`) is that it is a global variable that could get modified. Defining a static list of resources (similar to `urlpatterns`) seemed safer, since you can't accidentally modify the configuration.

Reply:
Ah, I think I faced a form of that problem, and added a flag which got flipped when the apps `.ready` method was called.